File: programmers/brute_force_search/숫자 야구.py
from itertools import permutations

def check(num: int, answer: list):
    '''
    부른 숫자와 정답을 받아와
    스트라이크/볼을 계산하는 함수
    '''
    strike = 0
    ball = 0
    
    for n, a in zip(map(int, str(num)), answer):
        if n == a:
            strike += 1
        elif n in answer:
            ball += 1

    return strike, ball

def my_solution(baseball):
    '''
    permutation으로 전체 순열을 만들고
    각 경기결과를 만족하지 않는 애들을 빼고
    남은 목록의 개수를 반환
    => set (bitwise 연산자: |, &, ... 사용) 활용
    '''
    candidates = set(permutations(range(1,9+1), 3))
    for num, strike, ball in baseball:
        remove_list = []
        for can in candidates:
            check_strike, check_ball = check(num, can)
            if not (strike == check_strike and ball == check_ball):
                remove_list.append(can)

        candidates -= set(remove_list)
            
    return len(candidates)

'''
another solution?
다른 사람 풀이 중에 각 경기결과를 바탕으로
(엄청나게 많은 and와 in , == 연산자 활용하여)
check 함수 사용 없이 리스트 컴프리 헨션으로
전체 가능 목록을 뽑아서 남은 목록과 & 연산으로 처리
'''

File: programmers/algorithm_data_structure/brute_force_search/숫자_야구.java
package algorithm_data_structure.brute_force_search;

import java.util.LinkedList;
import java.util.List;

public class 숫자_야구 {

    /*
     * 가능한 모든 수의 순열을 구하고
     * 직접 비교를 해보며 유효한 것만 남겨가는 방식
     * 
     * 순열을 구할 때 재귀함수를 사용
     * 
     * 아래와 같이 조합 수가 적을 땐 삼중 for문을 사용해도 된다.
     * 아래처럼 for문 3개에 if문 하거나
     * 
     * for (int i = 0 ...)
     *     for(int j = 0 ...)
     *         for(int k = 0 ...)
     *             if (i != j && j == k && i != k) {
     *                 ...
     *             }
     * 
     * 각각 i, j를 기준으로 하위 for문을 2개로 쪼개면 if문이 필요 없다.
     * (단, 코드 길이가 길어지고 중복이 많으며 다른 경우에 재사용이 어려움)
     */
    public int mySolution(int[][] baseball) {
        List<String> permutations = permutation();
        
        for(int[] info : baseball){
            String num = Integer.toString(info[0]);
            int strike = info[1];
            int ball = info[2];

            List<String> newPermutations = new LinkedList<>();

            for (String answer : permutations){
                int[] checkOutput = check(answer, num);

                if (checkOutput[0] == strike && checkOutput[1] == ball){
                    newPermutations.add(answer);
                }
            }

            permutations = newPermutations;
        }
        
        return permutations.size();
    }

    public int[] check(String answer, String number){
        int strike = 0;
        int ball = 0;

        for (int i = 0; i < number.length(); i++){
            // 값이 없으면 -1, 있으면 index 반환
            int idxInAnswer = answer.indexOf(number.charAt(i));
            if (idxInAnswer >= 0){
                if (idxInAnswer == i){
                    strike += 1;
                }else{
                    ball += 1;
                }
            }
        }

        return new int[]{strike, ball};
    }

    public List<String> permutation(){
        String range = "123456789";
        List<String> output = new LinkedList<>();
        
        permutation(range, "", output);
        return output;
    }

    public void permutation(String range, String prefix, List<String> output){
        if (prefix.length() == 3){
            output.add(prefix);
            return;
        }

        for (int i = 0; i < range.length(); i++){
            permutation(range.substring(0, i) + range.substring(i+1, range.length()),
                        prefix + range.charAt(i), output);
        }
    }
}